File: lis_nlogn.py
SIZE = 100
q = [ 0 ] * SIZE
p = [ 0 ] * SIZE
sol = [ 0 ] * SIZE
NR = 0
INF = 90000009
vec = [ 5,7,1,20,4,-6,-8,-9,2,12,10,11,8,0,30 ]
N = len(vec)
 
def insert(val, left, right):
 
    global NR
 
    m = (left + right ) // 2
 
    if(left == right):
 
        if right > NR:
            NR += 1
            q[ NR + 1 ] = INF
        q[left] = val
        return left
    else:
        if val <= q[m]:
            return insert(val, left, m)
        else:
            return insert(val, m + 1, right)
def main():
 
    for i in range(1, N):
        p[ i ] = insert(vec[i], 1, NR + 1)
 
    k = N - 1
    i = NR
    while(i != 0):
 
        while(p[k] != i):
           k -=1
        sol[i] = vec[k]
        i -= 1
    print("Longest Increasing Subsequence\nTime Complexity N Log N\nLength -> ", NR, "\nSubsequence -> ", end = " ")
    for i in range(1, NR+1):
        print(sol[i],end = " ")
    print("")
 
main()

File: lis_nlogn.cpp
#include <iostream>
#define FIN "scmax.in"
#define FOUT "scmax.out"
#define INF 2000000005
#define NMAX 100005

using namespace std;

int v[ NMAX ], N,
    p[ NMAX ], q[ NMAX ], s[ NMAX ], NR;

int insert(int val, int left, int right) {

    int m = (left + right) >> 1;

    if(left == right) {
             if(right > NR) {
                q[++NR + 1] = INF;
             }
             q[left] = val;
             return left;
    } else {

           if(val<=q[m]) {
              return insert(val, left, m);
           } else {
             return insert(val, m + 1, right);
           }
    }
}

void build() {

     int i;

     q[ 1 ] = INF;

     for(i = 1; i <= N; i++) {

         p[ i ] = insert(v[ i ], 1, NR + 1);
     }
}

void sol() {

     int i, k;

     for(k = N, i = NR; i; i--) {

            while(p[k] != i) k--;

            s[i] = v[k];
     }

     cout<<NR<<"\n";

     for(i = 1; i <= NR; i++) cout<<s[i]<<" ";

     cout<<"\n";
}

int main( int argc, char const *argv[] ) {

  freopen(FIN, "r", stdin);
  freopen(FOUT, "w", stdout);

  cin>>N;

  for(int i = 1; i <= N; ++i) {

      cin>>v[ i ];

  }

  build( );

  sol();

  return 0;
}
